synapse/handlers/presence.py

```diff
@@ -1249,6 +1308,27 @@ async def set_state(
 
         prev_state = await self.current_state_for_user(user_id)
 
+        # Always update the device specific information.
+        device_state = self.user_to_device_to_current_state.setdefault(
+            user_id, {}
+        ).setdefault(
+            device_id,
+            UserDevicePresenceState(
+                user_id,
+                device_id,
+                presence,
+                last_active_ts=self.clock.time_msec(),
+            ),
+        )
+        device_state.state = presence
+        device_state.last_active_ts = self.clock.time_msec()
```

Comment:
I would say have this be either the previous `UserPresenceState.last_active_ts` here, and `0` right above as the default(or the -1 you have in other places). If this call to `set_state()` is being provided with `online` as the state to set, it will be updated regardless(currently, every time). But if the state to set is something else, it should be pulled from the previous state. Whatever you do with it, don't use the current time as that would violate how `currently_active` is supposed to be used(and would fire in `handle_update()`.

Reply:
I think this code doesn't exist or got moved to a different PR, sorry. 😢